sail/src/main/java/org/apache/rya/rdftriplestore/inference/HasValueVisitor.java

```diff
@@ -0,0 +1,96 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+import org.apache.rya.api.RdfCloudTripleStoreConfiguration;
+import org.apache.rya.api.utils.NullableStatementImpl;
+import org.apache.rya.rdftriplestore.utils.FixedStatementPattern;
+import org.openrdf.model.Resource;
+import org.openrdf.model.URI;
+import org.openrdf.model.Value;
+import org.openrdf.model.vocabulary.OWL;
+import org.openrdf.model.vocabulary.RDF;
+import org.openrdf.query.algebra.StatementPattern;
+import org.openrdf.query.algebra.TupleExpr;
+import org.openrdf.query.algebra.Var;
+
+public class HasValueVisitor extends AbstractInferVisitor {
+    public HasValueVisitor(RdfCloudTripleStoreConfiguration conf, InferenceEngine inferenceEngine) {
+        super(conf, inferenceEngine);
+        include = true;
+    }
+
+    @Override
+    protected void meetSP(StatementPattern node) throws Exception {
+        final Var subjVar = node.getSubjectVar();
+        final Var predVar = node.getPredicateVar();
+        final Var objVar = node.getObjectVar();
+        // We can reason over two types of statement patterns:
+        // { ?var rdf:type :Restriction } and { ?var :property ?value }
+        // Both require defined predicate
+        if (predVar != null && predVar.getValue() != null) {
+            final URI predURI = (URI) predVar.getValue();
+            if (RDF.TYPE.equals(predURI) && objVar != null && objVar.getValue() != null
+                    && objVar.getValue() instanceof Resource) {
+                // If the predicate is rdf:type and the type is specified, check whether it can be
+                // inferred using any hasValue restriction(s)
+                final Resource objType = (Resource) objVar.getValue();
+                final Map<URI, Set<Value>> sufficientValues = inferenceEngine.getHasValueByType(objType);
+                if (sufficientValues.size() > 0) {
+                    final Var valueVar = new Var("v-" + UUID.randomUUID());
+                    TupleExpr currentNode = node.clone();
+                    for (URI property : sufficientValues.keySet()) {
+                        final Var propVar = new Var(property.toString(), property);
+                        final TupleExpr valueSP = new DoNotExpandSP(subjVar, propVar, valueVar);
+                        final FixedStatementPattern relevantValues = new FixedStatementPattern(objVar, propVar, valueVar);
+                        for (Value value : sufficientValues.get(property)) {
+                            relevantValues.statements.add(new NullableStatementImpl(objType, property, value));
+                        }
+                        currentNode = new InferUnion(currentNode, new InferJoin(relevantValues, valueSP));
```

Comment:
Okay.  I see.  I think my confusion here stems from my limited understanding of restrictions.  I was thinking of them as somehow being bound to a class or being a restriction placed on a class instead of as a class in themselves.  Thanks for clearing this up.